background/objectoriented.tex

```diff
@@ -1,210 +1,333 @@
+typedef struct {
+  uint32_t CTRL;
+  uint32_t CMD;
+  // ...
+  uint32_t SINGLECTRL;
+  // ...
+  uint32_t SINGLEDATA;
+  // ...
+} ADC;
+    \end{minted}
+    \caption{C}
+    \label{fig:back:adc-objects:c}
+  \end{subfigure}
+
+  \caption{Definition of ADC}
+  \label{list:back:adc-objects}
+\end{figure}
+
+\autoref{lst:back:adc-objects} shows how to define a Java class and Rust and C structs for the ADC on the Gecko.
+The memory layout of objects based on the definitions in \autoref{lst:back:adc-objects} are given in \label{fig:back:memlayout}.
+
 \begin{figure}[H]
 
   \centering
-  \subfloat[Java]{
+  \begin{subfigure}{0.31\textwidth}
     \begin{tabular}{|l|l|}
       \hline
-      \multicolumn{2}{|l|}{Sensor} \\
-      \hline \hline
-      0x0&Object tag \\
-      \hline
-      0x4&data\\
-      \hline
+      0x0&Object tag \\ \hline
+      0x4&CTRL       \\ \hline
+      0x8&CMD        \\ \hline
+      ...&...        \\ \hline
+      0x10&SINGLECTRL\\ \hline
+      ...&...        \\ \hline
+      0x28&SINGLEDATA\\ \hline
+      ...&...        \\ \hline
     \end{tabular}
+    \caption{Java}
     \label{fig:back:memlayout:java}
-  }
+  \end{subfigure}
   \hfill
-  \subfloat[Rust]{
+  \begin{subfigure}{0.31\textwidth}
     \begin{tabular}{|l|l|}
       \hline
-      \multicolumn{2}{|l|}{Sensor} \\
-      \hline \hline
-      0x0&data\\
-      \hline
+      0x0&CTRL       \\ \hline
+      0x4&CMD        \\ \hline
+      ...&...        \\ \hline
+      0xC&SINGLECTRL \\ \hline
+      ...&...        \\ \hline
+      0x24&SINGLEDATA\\ \hline
+      ...&...        \\ \hline
     \end{tabular}
+    \caption{Rust}
     \label{fig:back:memlayout:rust}
-  }
+  \end{subfigure}
   \hfill
-  \subfloat[C]{
+  \begin{subfigure}{0.31\textwidth}
     \begin{tabular}{|l|l|}
       \hline
-      \multicolumn{2}{|l|}{Sensor} \\
-      \hline \hline
-      0x0&data\\
-      \hline
+      0x0&CTRL       \\ \hline
+      0x4&CMD        \\ \hline
+      ...&...        \\ \hline
+      0xC&SINGLECTRL \\ \hline
+      ...&...        \\ \hline
+      0x24&SINGLEDATA\\ \hline
+      ...&...        \\ \hline
     \end{tabular}
+        \caption{C}
     \label{fig:back:memlayout:c}
-  }
+  \end{subfigure}
   \caption{Memory layout of objects}
   \label{fig:back:memlayout}
 
 \end{figure}
 
-And the control registers can be modified thru fields and member methods.
+By comparing \autoref{tab:back:memorymapped} and \autoref{fig:back:memlayout} we see that the memory layout of a struct defined in Rust and C has same exact layout as the memory mapped control register of the ADC.
+This suggests that if a pointer to the memory mapped I/O device is considered as a reference to an ADC object the object oriented pattern can be used to inter
+
+\subsection{Adding Object Functionality}
+
+This sections shows how we add functionality called methods to the MMIO objects.
+Both language uses static dispatch where the compiler knows the address of the function to apply at statically.
```

Comment:
language`s`

`to apply at statically` <- litt rar setning kanskje. hva betyr det?